File: Linked Lists/MergeKSortedLists.java
class Solution {
public:

struct mycmp{

 bool operator()(ListNode * a,ListNode * b)
 {
     return a->val > b->val;
 }
};

ListNode* mergeKLists(vector<ListNode*>& lists) {

 priority_queue<ListNode*,vector<ListNode *>,mycmp>pq;
 
 for(int i=0;i<lists.size();i++)
 {
     if(lists[i])
     {
         ListNode * ptr = lists[i];
         while(ptr != NULL)
         {
             pq.push(ptr);
             ptr = ptr->next;
         }
     }
     
 }
 
 ListNode * ptr = new ListNode(-1);
 ptr->next = NULL;
 ListNode * dummy = ptr;
 
 while(!pq.empty())
 {
     ListNode *temp = pq.top();
     ptr->next = temp;
     pq.pop();
     
     ptr = ptr->next;
 }
 ptr->next = NULL;
 
 return dummy->next;
}
};

File: Linked Lists/MergeKSortedLists.py
def mergeKLists(lists):
    
    def merge(lst1, lst2):
        dummy = pt = ListNode(-1)
        while lst1 and lst2:
            if lst1.val < lst2.val:
                pt.next = lst1
                lst1 = lst1.next
            else:
                pt.next = lst2
                lst2 = lst2.next
            pt = pt.next
            
        pt.next = lst1 if not lst2 else lst2
        return dummy.next
        
    
    if not lists:
        return None

    if len(lists) == 1:
        return lists[0]
    mid = len(lists)/2
    left = mergeKLists(lists[:mid])
    right = mergeKLists(lists[mid:])

    return merge(left, right)